rest/handler.go

```diff
@@ -35,62 +38,87 @@ func CreateHandler(ctx context.Context, prometheusURL string) (http.Handler, err
-	var subHandler func(w http.ResponseWriter, r *http.Request, s langserver.HeadlessServer, requestID string)
-
-	requestID := fmt.Sprint(atomic.AddInt64(&h.requestCounter, 1), ".promql")
-
-	switch r.URL.Path {
-	case "/diagnostics":
-		subHandler = diagnosticsHandler
-	case "/hover":
-		subHandler = hoverHandler
-	case "/completion":
-		subHandler = completionHandler
-	case "/signatureHelp":
-		subHandler = signatureHelpHandler
-	default:
-		http.NotFound(w, r)
-		return
+// CreateInstHandler creates an instrumented http.Handler for the PromQL langserver REST API.
+//
+// Expects the URL of a Prometheus server as the second argument and a Registry as third argument.
+func CreateInstHandler(ctx context.Context, prometheusURL string, r *prometheus.Registry) (http.Handler, error) {
+	langserver, err := langserver.CreateHeadlessServer(ctx, prometheusURL)
+	if err != nil {
+		return nil, err
 	}
 
-	exprs, ok := r.URL.Query()["expr"]
+	ls := &langserverHandler{langserver: langserver}
+	ls.m = make(map[string]http.Handler)
+	ls.createDefaultInstHandlers(r)
 
-	if !ok || len(exprs) == 0 {
-		http.Error(w, "Param expr is not specified", 400)
-		return
+	return ls, nil
+}
+
+func (h *langserverHandler) createDefaultHandlers() {
+	h.Handle("/diagnostics", newURLHandler("/diagnostics", h, diagnosticsHandler))
+	h.Handle("/completion", newURLHandler("/completion", h, completionHandler))
+	h.Handle("/hover", newURLHandler("/hover", h, hoverHandler))
+	h.Handle("/signatureHelp", newURLHandler("/signatureHelp", h, signatureHelpHandler))
+}
+
+func (h *langserverHandler) createDefaultInstHandlers(r *prometheus.Registry) {
+	httpRequestsTotal := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "http_requests_total",
+		Help: "Count of all HTTP requests",
+	}, []string{"code", "method"})
+
+	httpRequestDuration := prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "http_request_duration_seconds",
+		Help: "Duration of all HTTP requests base on endpoint being queried",
+	}, []string{"code", "handler"})
+
+	r.MustRegister(httpRequestsTotal)
+	r.MustRegister(httpRequestDuration)
+
+	instrumentFunc := func(name string, h http.Handler) http.Handler {
+		return promhttp.InstrumentHandlerDuration(
+			httpRequestDuration.MustCurryWith(prometheus.Labels{"handler": name}),
+			promhttp.InstrumentHandlerCounter(httpRequestsTotal, h))
 	}
 
-	defer func() {
-		h.langserver.DidClose(r.Context(), &protocol.DidCloseTextDocumentParams{
-			TextDocument: protocol.TextDocumentIdentifier{
-				URI: requestID,
-			},
-		},
-		)
-	}()
+	diagnostics := instrumentFunc("diagnostics", newURLHandler("/diagnostics", h, diagnosticsHandler))
+	completion := instrumentFunc("completion", newURLHandler("/completion", h, completionHandler))
+	hover := instrumentFunc("hover", newURLHandler("/hover", h, hoverHandler))
+	signatureHelp := instrumentFunc("signatureHelp", newURLHandler("/signatureHelp", h, signatureHelpHandler))
+	//we don't want to instrument the metrics endpoint
```

Comment:
Why not?

Reply:
for instance the number of http request to the metrics endpoint would be counted along with other http request to endpoint like /hover /completion  etc. since we most likely want to monitor this endpoints including the metrics endpoint for instrumentation would only include unwanted metrics as we probably don't care much about the metrics endpoint.

Reply:
Some times monitoring the `/metrics` endpoint actually produces interesting results: https://github.com/prometheus/prometheus/issues/6785

Reply:
fixed